### PR TITLE
Instrument Results Importer/Parser improvements and fixtures

### PR DESCRIPTION
**Pull request by ramonski**
## Description of the issue/feature this PR addresses

This PR tries to untangle the instrument import logic baked into SENAITE to make it

1. Understandable
2. Python 3 compatible
3. Better extendible

## Current behavior before PR

Instrument import functionality is a historical grown maze

## Desired behavior after PR is merged

Instrument import functionality becomes maintainable

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: src/senaite/core/exportimport/dataimport.py [hunk 1]
```diff
@@ -99,27 +100,34 @@ def getProjectName(self):
...
-        for item in brains:
-            items.append((item.UID, item.Title))
+        """Get all instruments that have at least one import interface selected
+        """
+        setup_catalog = api.get_tool(SETUP_CATALOG)
+        brains = setup_catalog(portal_type="Instrument", is_active=True)
+        items = [("", _("... Choose an Instrument ..."))]
+        for brain in brains:
+            instrument = api.get_object(brain)
+            if not instrument.getImportDataInterface():
```
Comment (xispa) [inline src/senaite/core/exportimport/dataimport.py]: For some reason, `ImportDataInterface` is a multivalued `StringField` :thinking: that relies on a vocabulary: https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/content/instrument.py#L223-L225


Thus, this condition does not work as expected:

```python
>>> instrument.getImportDataInterface()
['']
```

File: src/senaite/core/exportimport/instruments/importer.py [hunk 2]
```diff
@@ -0,0 +1,1049 @@
...
+        :returns: Converted analysis result
+        """
+        result_options = analysis.getResultOptions()
+
+        if result_options:
+            # Handle result options as integer values
+            result_values = map(
+                lambda r: r.get("ResultValue"), result_options)
+            # check if the result is in result options
+            if "{:.0f}".format(result) in result_values:
```
Comment (xispa) [inline src/senaite/core/exportimport/instruments/importer.py]: Not sure about this. There is no type forced (`subfield_types`) for the subfield `ResultValue` from [`ResultOptions` field](https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/content/abstractbaseanalysis.py#L613), but the [validator `result_options_value_validator`](https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/validators.py#L624). And the validator accepts `floats`... Thus, if a `ResultValue` comes to be e.g. `2.5`, the system will wrongly resolve to `2`:

```python
>>> "{:.0f}".format(2.5)
'2'
```
I would keep this as much *naive* as possible.

Also, multi-select and multi-choice values are not taken into account. I would rely on [`ResultType`](https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/content/abstractbaseanalysis.py#L598) to know what type of result we expect and then return back the right value:

- If numeric, text or string, return the result without processing
- If select, return the `ResultValue` (without int conversion)
- If multiselect or multichoice, return a list of `ResultValue` (without int conversion)

File: src/senaite/core/exportimport/instruments/importer.py [hunk 3]
```diff
@@ -0,0 +1,1049 @@
...
+                continue
+            # remember the valid service keyword
+            keywords.append(keyword)
+
+        if len(keywords) == 0:
+            self.warn(_("No services could be found for parsed keywords"))
+
+        return keywords
+
+    def is_valid_keyword(self, keyword):
```
Comment (xispa) [inline src/senaite/core/exportimport/instruments/importer.py:317]: I think is worth to memoize this function

File: src/senaite/core/exportimport/instruments/importer.py [hunk 4]
```diff
@@ -0,0 +1,1049 @@
...
+        infile = self.parser.getInputFile()
+
+        ancount = 0
+        updated_analyses = []
+        importedinsts = {}
+        importedars = {}
+
+        for sid, results in parsed_results.items():
+            # get the sample
+            sample = self.get_sample_by_id(sid)
```
Comment (xispa) [inline src/senaite/core/exportimport/instruments/importer.py]: This line (and the function `get_sample_by_id` as well) can be removed cause the `sample` variable is never used in this function. System relies on `get_analyses_for` function to directly get the analyses, without the need of looking for the sample first.

File: src/senaite/core/exportimport/instruments/importer.py [hunk 5]
```diff
@@ -0,0 +1,1049 @@
...
+                                    self.allowed_sample_states_msg),
+                                "sid": sid}))
+                continue
+
+            # we have an instrument
+            elif len(analyses) == 0 and self.instrument:
+                # Create a new ReferenceAnalysis and link it to the Instrument.
+                refsample = self.get_reference_sample_by_id(sid)
+                if not refsample:
+                    self.warn(_("No Reference Sample found for ${sid}",
```
Comment (xispa) [inline src/senaite/core/exportimport/instruments/importer.py]: Replace by "No Sample found for ${sid}" (remove "Reference"), cause this condition will be true also when no *routine* analyses (with or without sample) are found, that will probably happen more often

File: src/senaite/core/exportimport/instruments/importer.py [hunk 6]
```diff
@@ -0,0 +1,1049 @@
...
+            fn_attachments[fn].append(att)
+        return fn_attachments
+
+    def is_analysis_allowed(self, analysis):
+        """Filter analyses that match the import criteria
+        """
+        portal_type = api.get_portal_type(analysis)
+        if portal_type != "Analysis":
+            # Reference Analysis / Duplicate Analysis
+            return True
```
Comment (xispa) [inline src/senaite/core/exportimport/instruments/importer.py:889]: I'd say this is used to allow the update of `ReferenceAnalysis` objects even when they were already verified, so if the new result is within ranges (while previous not), the instrument would automatically be considered as valid. This probably deserves a thorough discussion in another PR, but for now I would at least keep "DuplicateAnalyses" out of the equation. So I would replace this condition as follows:

```python
if IReferenceAnalysis.providedBy(analysis):
    return True
```

**Review by xispa — CHANGES_REQUESTED** [review]
Man, thanks for applying some order to this results import mess. I've suggested some changes, let me know what you think

File: src/senaite/core/exportimport/dataimport.py (hunk 1 above)
Comment (ramonski) [reply src/senaite/core/exportimport/dataimport.py]: Thanks for finding this quirk! Fixed in 35664181f

File: src/senaite/core/exportimport/instruments/importer.py (hunk 3 above)
Comment (ramonski) [reply src/senaite/core/exportimport/instruments/importer.py:317]: Indeed, makes sense! Added the `memoize_contextless` decorator in 824fb6bc1

File: src/senaite/core/exportimport/instruments/importer.py (hunk 4 above)
Comment (ramonski) [reply src/senaite/core/exportimport/instruments/importer.py]: Good catch! Removed here: e37297118

File: src/senaite/core/exportimport/instruments/importer.py (hunk 5 above)
Comment (ramonski) [reply src/senaite/core/exportimport/instruments/importer.py]: Updated the message in 9b1f2c04d

File: src/senaite/core/exportimport/instruments/importer.py (hunk 6 above)
Comment (ramonski) [reply src/senaite/core/exportimport/instruments/importer.py:889]: Refactored in d0ec1c2ed

File: src/senaite/core/exportimport/instruments/importer.py (hunk 2 above)
Comment (ramonski) [reply src/senaite/core/exportimport/instruments/importer.py]: Wow, good catch!

Indeed, when the `ResultType` is set to "Selection List", we can assign both `int` and `float` values for the results:
Result Options

However, I guess most of the time people use `int` values, e.g. `0` for `negative` and `1` for `positive`.

Nevertheless, I adapted now the code here for Selection Lists: 9f729b1c8

Not sure if and how instruments can import multi results for the other fields. Maybe this rather be implemented in a custom instrument importer?

Comment (xispa) [reply src/senaite/core/exportimport/instruments/importer.py]: > Not sure if and how instruments can import multi results for the other fields. Maybe this rather be implemented in a custom instrument importer?

I haven't stumbled upon this requirement (yet), so I think import of multi-results for interim fields from instrument results files might be an edge case at the moment. So I think for now is easier to use a custom instrument importer for this use case

**Review by xispa — APPROVED** [review]
Thanks @ramonski :1st_place_medal:

**Pull request merged**